keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,164 @@
+
+## Proposal
+Use memory controller interfaces of cgroups v2 to support memory qos of pod qos class.
+
+Kubernetes cgroup manager uses `memory.limit_in_bytes` in cgroups v1 to limit memory capacity and uses `oom_scores` to determine order of killing container process when OOM occurs. 
+
+Above implement has following disadvantages:
+- `Guaranteed` memory can not be fully reserved, page cache is at risk of being recycled
+- `Burstable` memory allocation overcommit ( request < limit ) could increase OOM possibility when kernel memory runs out
+
+cgroups v2 introduces a better way to limit and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high    | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup’s memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a “soft guarantee” that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup’s memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. |
+
+This proposal uses `memory.min` to retain request / limit memory for `Guaranteed` Pod. And for `Burstable` Pod, it uses `memory.low` to achieve best-effort memory protection as well as using `memory.high` to throttle allocation of memory overcommit.
+
+### User Stories (Optional)
+- Workload is memory-sensitive, this feature can retain memory to reduce allocation latency
+- Workload is overcommitted in memory, this feature can throttle memory overcommit to increase stability by reducing risk of kernel OOM
+
+### Risks and Mitigations
+- `Guaranteed` memory reservation could reduce node memory supply
+- `Burstable` memory overcommit allocation could slow down in condition
+
+## Design Details
+### Feature Gate
+To set `--feature-gates=MemoryQOS=true` to enable the feature.
+
+### CGroup Resource
+Add a new `Unified` variable for `memory.min / low / high` and for extension of other parameters in cgroups v2 memory controller. Suggest same meaning in OCI runtime specs https://github.com/opencontainers/runtime-spec/pull/1040
```

Comment:
Add to `ResourceConfig` (internal kubelet resource representation struct).

Reply:
Done.